File: fire_response_web/app/Models/FireReports.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class FireReports extends Model
{
    /** @use HasFactory<\Database\Factories\FireReportsFactory> */
    use HasFactory;

    protected $fillable = [
        'reported_by',
        'location',
        'landmark',
        'description',
        'status',
        'latitude',
        'longitude',
        'contact_info',
        'fireStationId',
        'marked_as_contained_by_id',
        'marked_as_contained_at',
        'barangay_id',
        'marked_as_false_alarm_by',
        'marked_as_false_alarm_at',
        'false_alarm_image',
        'fire_report_image', 

    ];

    protected $casts = [
        'marked_as_contained_at' => 'datetime',
        'marked_as_false_alarm_at' => 'datetime',
    ];
    

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function fireReports()
    {
        return $this->hasMany(FireReports::class, 'reported_by');
    }

    public function fireStation()
    {
        return $this->belongsTo(FireStation::class, 'fireStationId');
    }

    public function assignedTeams() {
        return $this->hasMany(Team::class, 'assignedFireIncident');
    }

    public function stages() {
        return $this->belongsTo(RealTimeFireReport::class, 'fire_report_id');
    }

    public function markedBy() {
        return $this->belongsTo(Firefighter::class, 'marked_as_contained_by_id');
    }

    public function firefighterReports() {
        return $this->hasMany(FirefighterReports::class, 'fireReportId');
    }

    public function responders() {
        return $this->belongsToMany(Firefighter::class, 'firefighter_reports', 'fireReportId', 'fireFighterId')
        ->with('user:id,userFirstName,userLastName');
    }

    public function barangay() {
        return $this->belongsTo(Barangay::class);
    }

    public function markedAsFalseAlarmBy() {
        return $this->belongsTo(User::class, 'marked_as_false_alarm_by');
    }

}

File: fire_response_app/lib/models/fire_reports.dart
class FireReport {
  final int id;
  final int reportedBy;
  final int? fireStationId;
  final String location;
  final double? latitude;
  final double? longitude;
  final String? landmark;
  final String? description;
  final String? contactInfo;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int? markedAsContainedById;
  final DateTime? markedAsContainedAt;

  FireReport({
    required this.id,
    required this.reportedBy,
    this.fireStationId,
    required this.location,
    this.latitude,
    this.longitude,
    this.landmark,
    this.description,
    this.contactInfo,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.markedAsContainedById,
    this.markedAsContainedAt,
  });

  factory FireReport.fromJson(Map<String, dynamic> json) {
    return FireReport(
      id: json['id'],
      reportedBy: json['reported_by'],
      fireStationId:
          json['fireStationId'] != null ? json['fireStationId'] as int : null,
      location: json['location'],
      latitude:
          json['latitude'] != null
              ? (json['latitude'] is String
                  ? double.tryParse(json['latitude']) ?? 0.0
                  : json['latitude'])
              : null,
      longitude:
          json['longitude'] != null
              ? (json['longitude'] is String
                  ? double.tryParse(json['longitude']) ?? 0.0
                  : json['longitude'])
              : null,
      landmark: json['landmark'],
      description: json['description'],
      contactInfo: json['contact_info'],
      status: json['status'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      markedAsContainedById:
          json['marked_as_contained_by_id'] != null
              ? json['marked_as_contained_by_id'] as int
              : null,
      markedAsContainedAt:
          json['marked_as_contained_at'] != null
              ? DateTime.parse(json['marked_as_contained_at'])
              : null,
    );
  }
}
